kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/PaginatedScanSuite.scala

```diff
@@ -0,0 +1,268 @@
+    with ExpressionTestUtils with SQLHelper with DeltaTableWriteSuiteBase {
+
+  private val logger = LoggerFactory.getLogger(classOf[PaginatedScanSuite])
+  // TODO: test call hasNext() twice
+
+  test("Read First Page: basic pagination with single JSON file") {
+    withTempDir { tempDir =>
+      /**
+       *   Creates a Delta table with 10 Parquet data files; each file contains 10 rows
+       * * 00000000000000000000.json contains 10 AddFile actions — one per file
+       */
+      spark.range(0, 100, 1, 10).write.format("delta").save(tempDir.getCanonicalPath)
+
+      val snapshot = latestSnapshot(tempDir.getCanonicalPath)
+      val pageSize = 8 // TODO: change different page sizes
+      val paginatedScan = snapshot.getScanBuilder
+        .buildPaginated(pageSize, Optional.empty() /* page token opt */ )
+        .asInstanceOf[PaginatedScanImpl]
+
+      /**
+       * Creates a custom engine with batch size 5;
+       * 00000000000000000000.json contains 3 batches, 10 active AddFiles in total
+       * Batch 1: 5 rows, 2 selected AddFiles
+       * Batch 2: 5 rows, 5 selected AddFiles
+       * Batch 3: 3 rows, 3 selected AddFiles
+       */
+      val hadoopConf = new org.apache.hadoop.conf.Configuration()
+      hadoopConf.set("delta.kernel.default.json.reader.batch-size", "5")
+      val customEngine = DefaultEngine.create(hadoopConf)
+
+      val paginatedIter = paginatedScan.getScanFiles(customEngine)
+      val firstPageFiles = collection.mutable.Buffer[FilteredColumnarBatch]()
+      while (paginatedIter.hasNext) {
+        val batch = paginatedIter.next()
+        firstPageFiles += batch
+      }
+      assert(firstPageFiles.nonEmpty, "First page should contain some files")
+
+      /** Verify we get 3 batches and 10 ScanFiles in total in the first page */
+      val fileCounts: Seq[Long] = firstPageFiles.map(_.getPreComputedNumSelectedRows.get().toLong)
+      val totalFileCountsReturned = fileCounts.sum
+      assert(
+        totalFileCountsReturned == 10,
+        s"First page should contain 10 files, got $totalFileCountsReturned")
+
+      assert(
+        fileCounts.length == 3,
+        s"First page should contain 3 batches, got ${fileCounts.length}")
+
+      logger.info(s"Total num batches returned in page one = ${fileCounts.length}")
+      logger.info(s"Total num Parquet Files fetched in page one = $totalFileCountsReturned")
+
+      /** Verify we get correct page token for next page */
+      val lastReadLogFileName = PageToken.fromRow(paginatedIter.getCurrentPageToken)
+        .getLastReadLogFileName
+      val lastReturnedRowIndex = PageToken.fromRow(paginatedIter.getCurrentPageToken)
+        .getLastReturnedRowIndex
+
+      assert(
+        lastReadLogFileName.endsWith("00000000000000000000.json"),
+        s"New Page Token: LastReadLogFilename should be 00000000000000000000.json, " +
+          s"got $lastReadLogFileName")
+
+      assert(
+        lastReturnedRowIndex == 12,
```

Comment:
to me, `12` is a magic number. where did this come from? what does it mean? if we change other parts of our code, and this assert starts to fail, how do we know what this value should be updated to?

https://en.wikipedia.org/wiki/Magic_number_(programming)